freeipa/src/main/java/com/sequenceiq/freeipa/service/stack/FreeIpaHealthDetailsService.java

```diff
@@ -53,35 +57,69 @@
 
     public HealthDetailsFreeIpaResponse getHealthDetails(String environmentCrn, String accountId) {
         Stack stack = stackService.getByEnvironmentCrnAndAccountIdWithLists(environmentCrn, accountId);
-        InstanceMetaData master = findMaster(stack);
-        Optional<RPCResponse<Boolean>> rpcResponse = Optional.empty();
-        try {
-            rpcResponse = Optional.ofNullable(checkFreeIpaHealth(stack, master));
-        } catch (FreeIpaClientException e) {
-            LOGGER.error("Unable to check the health of FreeIPA.", e);
+        List<InstanceMetaData> instances = stack.getInstanceGroups().stream()
+                .flatMap(instanceGroup -> instanceGroup.getInstanceMetaData().stream()).collect(Collectors.toList());
+        HealthDetailsFreeIpaResponse response = new HealthDetailsFreeIpaResponse();
+        // The code here is weird because the server_conncheck API requires a specific node
+        // and we don't know what determines what that node needs to be.  But if we use
+        // the wrong master, we get an exception that says
+        //   invalid 'cn': must be "ipaserver0.xx.wl.cloudera.site"
+        // so we parse that, set that as master, and then retry the node.
+        String masterCn = findMaster(stack).getDiscoveryFQDN();
+        int retries = 0;
+        for (int idx = 0; idx < instances.size(); idx++) {
```

Comment:
you should use `for (InstanceMetaData instance : instances)` as you don't use the index elsewhere

Reply:
I cannot update the list of instances if using this style for loop.  Per next comment, I add it back to the list to retry since I cannot idx--.

Reply:
changed.